apps/prairielearn/src/batched-migrations/20250215015626_course_instance_usages__workspaces__backfill.ts

```diff
@@ -0,0 +1,26 @@
+import { z } from 'zod';
+
+import { makeBatchedMigration } from '@prairielearn/migrations';
+import { loadSqlEquiv, queryAsync, queryRow } from '@prairielearn/postgres';
+
+const sql = loadSqlEquiv(import.meta.url);
+
+const CUTOFF_DATE = '2025-02-15T00:00:00Z';
+
+export default makeBatchedMigration({
+  async getParameters() {
+    // First delete usage data older than a hard-coded date
+    await queryAsync(sql.delete_old_usages, { CUTOFF_DATE });
+
+    // Only backfill from workspaces that are older than the cutoff date
+    const max = await queryRow(
+      sql.select_bounds,
+      { CUTOFF_DATE },
+      z.bigint({ coerce: true }).nullable(),
+    );
+    return { min: 1n, max, batchSize: 10_000 };
```

Comment:
This takes about 7 seconds per batch so the whole thing should be done in less than an hour.

Weirdly, using a batch size of 100k caused the query time to exceed 30 minutes (I killed it). It must be generating a really bad plan for some reason.

Reply:
Huh! That's really weird. Glad this'll run pretty fast in practice though.